File: Graph_Functions/DFS_Iterative/dfs_iter.cpp
#include <bits/stdc++.h>
using namespace std;

// Iterative DFS function that takes the adjacency list and an empty container
void dfs(const vector<vector<int>>& adj, vector<int>& dfs_order) {
    int n = adj.size();
    
    // Create a visited vector to track which nodes have been visited
    vector<bool> visited(n, false);
    
    // Stack for the iterative DFS
    stack<int> st;
    
    // Let's start DFS from node 0 (or any other node)
    st.push(0);
    visited[0] = true;
    
    while (!st.empty()) {
        int node = st.top();
        st.pop();
        
        // Add the current node to the DFS order
        dfs_order.push_back(node);
        
        // Explore all neighbors of the current node
        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                st.push(neighbor);
            }
        }
    }
}

int main() {
    // Define the graph as an adjacency list
    int n = 5;  // Number of nodes
    vector<vector<int>> adj(n);
    
    // Adding edges (undirected graph)
    adj[0].push_back(1);
    adj[0].push_back(2);
    adj[1].push_back(0);
    adj[2].push_back(0);
    adj[2].push_back(3);
    adj[3].push_back(2);
    adj[3].push_back(4);
    adj[4].push_back(3);
    
    // Container to store the DFS order
    vector<int> dfs_order;
    
    // Call the DFS function
    dfs(adj, dfs_order);
    
    // Print the DFS order
    cout << "DFS Order: ";
    for (int node : dfs_order) {
        cout << node << " ";
    }
    cout << endl;
    
    return 0;
}

File: Graph_Functions/DFS_Iterative/dfs_iter.java
package DFS_Iterative;
import java.util.*;

public class dfs_iter {

    // Iterative DFS function that takes the adjacency list and an empty container
    public static void dfs(List<List<Integer>> adj, List<Integer> dfsOrder) {
        int n = adj.size();
        
        // Create a visited array to track which nodes have been visited
        boolean[] visited = new boolean[n];
        
        // Stack for iterative DFS
        Stack<Integer> stack = new Stack<>();
        
        // Let's start DFS from node 0 (or any other node)
        stack.push(0);
        visited[0] = true;
        
        while (!stack.isEmpty()) {
            int node = stack.pop();
            
            // Add the current node to the DFS order
            dfsOrder.add(node);
            
            // Explore all neighbors of the current node
            for (int neighbor : adj.get(node)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    stack.push(neighbor);
                }
            }
        }
    }

    // Main function where the DFS function is called and use case is demonstrated
    public static void func_body() {
        // Define the graph as an adjacency list
        int n = 5;  // Number of nodes
        List<List<Integer>> adj = new ArrayList<>();
        
        // Initialize adjacency list
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        
        // Adding edges (undirected graph)
        adj.get(0).add(1);
        adj.get(0).add(2);
        adj.get(1).add(0);
        adj.get(2).add(0);
        adj.get(2).add(3);
        adj.get(3).add(2);
        adj.get(3).add(4);
        adj.get(4).add(3);
        
        // Container to store the DFS order
        List<Integer> dfsOrder = new ArrayList<>();
        
        // Call the DFS function
        dfs(adj, dfsOrder);
        
        // Print the DFS order
        System.out.print("DFS Order: ");
        for (int node : dfsOrder) {
            System.out.print(node + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        // Main function where the DFS function is called
        func_body();
    }
}
